### ISSUE TITLE
how much memory needed for 2x150bp Illumina 30x mammalian genome assembly?

### ISSUE THREAD

**Issue by avilella** [ablab/spades#456]
Hi, I am running spades on a ~30x 2x150bp Illumina NextSeq dataset for a mammalian size genome with the options below. How much memory is needed for this for a human-sized mammalian genome?

```
/spades.py  -k      21,33,55,77     --careful       -1 SRR8377759_1.fastq.gz  -2 SRR8377759_2.fastq.gz -o spades_output
```

This is running on an AWS Ubuntu Server 18.04 instance with 95Gb of RAM, but I also created extra swap partition of a total of about 300Gb.

```
== Running: /home/ubuntu/SPAdes-3.14.0-Linux/bin/spades-hammer /home/ubuntu/data/public_data/cat/lilbub/spades_output/corrected/configs/config.info

  0:00:00.000     8M / 14M   INFO    General                 (main.cpp                  :  75)   Starting BayesHammer, built from refs/heads/spades_3.14.0, git revision c831f9be30a4364383cd4ebe5b78cdfcaad1acc9
  0:00:00.003     8M / 14M   INFO    General                 (main.cpp                  :  76)   Loading config from /home/ubuntu/data/public_data/cat/lilbub/spades_output/corrected/configs/config.info
  0:00:00.006     8M / 14M   INFO    General                 (main.cpp                  :  78)   Maximum # of threads to use (adjusted due to OMP capabilities): 16
  0:00:00.006     8M / 14M   INFO    General                 (memory_limit.cpp          :  49)   Memory limit set to 92 Gb
  0:00:00.006     8M / 14M   INFO    General                 (main.cpp                  :  86)   Trying to determine PHRED offset
  0:00:00.006     8M / 14M   INFO    General                 (main.cpp                  :  92)   Determined value is 33
  0:00:00.006     8M / 14M   INFO    General                 (hammer_tools.cpp          :  38)   Hamming graph threshold tau=1, k=21, subkmer positions = [ 0 10 ]
  0:00:00.006     8M / 14M   INFO    General                 (main.cpp                  : 113)   Size of aux. kmer data 24 bytes
     === ITERATION 0 begins ===
  0:00:00.007     8M / 14M   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index
  0:00:00.007     8M / 14M   INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 256 files using 16 threads. This might take a while.
  0:00:00.007     8M / 14M   INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 1024
  0:00:00.007     8M / 14M   INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 1.91665 Gb
  0:00:00.007     8M / 14M   INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 262144
  0:00:00.017    11G / 11G   INFO   K-mer Splitting          (kmer_data.cpp             :  97)   Processing /home/ubuntu/data/public_data/cat/lilbub/SRR8377759_1.fastq.gz
  0:00:16.926    11G / 11G   INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 3181193 reads
  0:00:35.683    11G / 11G   INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 6316812 reads
  0:00:52.595    11G / 11G   INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 9479364 reads
  0:01:12.103    11G / 11G   INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 12916359 reads
  0:01:30.396    11G / 11G   INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 16055033 reads
  0:01:49.253    11G / 11G   INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 19255361 reads
  0:02:08.359    11G / 11G   INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 22516502 reads
  0:02:28.931    11G / 11G   INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 25978697 reads
  0:02:48.185    11G / 11G   INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 29147039 reads
  0:03:07.517    11G / 11G   INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 32312726 reads
  0:03:28.820    11G / 11G   INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 35509786 reads
  0:07:05.024    11G / 11G   INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 67918747 reads
  0:14:41.278    11G / 11G   INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 136563994 reads
  0:29:35.330    11G / 11G   INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 270796598 reads
  0:33:57.772    11G / 11G   INFO   K-mer Splitting          (kmer_data.cpp             :  97)   Processing /home/ubuntu/data/public_data/cat/lilbub/SRR8377759_2.fastq.gz
  0:59:34.494    11G / 11G   INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 539853684 reads
  1:08:36.359    11G / 11G   INFO   K-mer Splitting          (kmer_data.cpp             : 112)   Total 620724532 reads processed
  1:08:36.690   178M / 8G    INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.
  2:30:14.507    36M / 61G   INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 11285155542 kmers in total.
  2:30:14.507    36M / 61G   INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.
  2:40:36.693    36M / 61G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices
<jemalloc>: Error in malloc(): out of memory
 Requested: 338542416


== Error ==  system call for: "['/home/ubuntu/SPAdes-3.14.0-Linux/bin/spades-hammer', '/home/ubuntu/data/public_data/cat/lilbub/spades_output/corrected/configs/config.info']" finished abnormally, OS return value: -6
```

**Comment by asl** [comment]
Hello

Thank you for your interest in SPAdes.

It's hard to predict as the memory consumption depends on many factors including the genome size, error rate, repeat content, heterozygosity rate among the others. Typically 100-200 Gb of free RAM should be enough though

**Comment by avilella** [comment]
Hi again,

Just to update, I attempted the assembly again with `--careful --only_assemble` options in a 512Gb Ubuntu 18.04 AWS instance. It successfully did the K21, K33 steps, but crashed with an out of memory message at the K55 step:



```
===== K55 started.                                                                                                                                                             
                                                                                                                              
                                                                                                                                  
== Running: /home/ubuntu/SPAdes-3.14.0-Linux/bin/spades-core /home/ubuntu/data/public_data/cat/lilbub/spades_output/K55/configs/config.info /home/ubuntu/data/public_data/cat/lilbub/spades_output/K55/configs/careful_mode.info
                                                                                                                                                            
  0:00:00.000    10M / 13M   INFO    General                 (main.cpp                  :  73)   Loaded config from /home/ubuntu/data/public_data/cat/lilbub/spades_output/K55/configs/config.info
  0:00:00.001    10M / 13M   INFO    General                 (main.cpp                  :  73)   Loaded config from /home/ubuntu/data/public_data/cat/lilbub/spades_output/K55/configs/careful_mode.info
  0:00:00.001    10M / 13M   INFO    General                 (memory_limit.cpp          :  49)   Memory limit set to 500 Gb                    
  0:00:00.001    10M / 13M   INFO    General                 (main.cpp                  :  83)   Starting SPAdes, built from refs/heads/spades_3.14.0, git revision c831f9be30a4364383cd4ebe5b78cdfcaad1acc9
  0:00:00.001    10M / 13M   INFO    General                 (main.cpp                  :  84)   Maximum k-mer length: 128                     
  0:00:00.001    10M / 13M   INFO    General                 (main.cpp                  :  85)   Assembling dataset (/home/ubuntu/data/public_data/cat/lilbub/spades_output/dataset.info) with K=55                                                                             

[...]

  7:35:56.488   169G / 224G  INFO    General                 (simplification.cpp        : 388)   PROCEDURE == Simplification cycle, iteration 13
  7:35:56.488   169G / 224G  INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper
  7:35:56.489   169G / 224G  INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times
  7:35:56.489   169G / 224G  INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover
  7:35:56.489   169G / 224G  INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times
  7:35:56.490   169G / 224G  INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover
  7:35:56.490   169G / 224G  INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times
  7:36:03.261   169G / 224G  INFO   StageManager             (stage.cpp                 : 166)   STAGE == Gap Closer
  7:36:03.262   169G / 224G  INFO    General                 (graph_pack.hpp            : 105)   Index refill
  7:36:03.262   169G / 224G  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index
  7:36:03.262   169G / 224G  INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 256 files using 16 threads. This might take a while.
  7:36:03.263   169G / 224G  INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 1024
  7:36:03.264   169G / 224G  INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 6.98534 Gb
  7:36:03.264   169G / 224G  INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 131072
  7:36:14.971   179G / 224G  INFO    General                 (edge_index_builders.hpp   :  77)   Processed 7980492 edges
  7:36:26.319   179G / 224G  INFO    General                 (edge_index_builders.hpp   :  77)   Processed 15945277 edges
  7:36:36.202   179G / 224G  INFO    General                 (edge_index_builders.hpp   :  77)   Processed 23912787 edges
  7:36:46.110   179G / 224G  INFO    General                 (edge_index_builders.hpp   :  77)   Processed 31879531 edges
  7:36:55.711   179G / 224G  INFO    General                 (edge_index_builders.hpp   :  77)   Processed 39835574 edges
  7:37:05.374   179G / 224G  INFO    General                 (edge_index_builders.hpp   :  77)   Processed 47817408 edges
  7:37:15.350   179G / 224G  INFO    General                 (edge_index_builders.hpp   :  77)   Processed 55778723 edges
  7:37:24.162   179G / 224G  INFO    General                 (edge_index_builders.hpp   :  77)   Processed 62707086 edges
  7:37:24.163   179G / 224G  INFO    General                 (edge_index_builders.hpp   :  82)   Used 62707086 sequences.
  7:37:24.659   169G / 224G  INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.
  7:38:27.991   169G / 224G  INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 4056378073 kmers in total.
  7:38:27.991   169G / 224G  INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.
  7:39:59.931   169G / 224G  INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices
  7:39:59.932   170G / 224G  ERROR   General                 (mmapped_reader.hpp        :  52)   mmap(2) failed. Reason: Cannot allocate memory. Error code: 12
  7:39:59.932   170G / 224G  ERROR   General                 (mmapped_reader.hpp        :  52)   mmap(2) failed. Reason: Cannot allocate memory. Error code: 12
  7:39:59.932   170G / 224G  ERROR   General                 (mmapped_reader.hpp        :  52)   mmap(2) failed. Reason: Cannot allocate memory. Error code: 12
  7:39:59.932   170G / 224G  ERROR   General                 (mmapped_reader.hpp        :  52)   mmap(2) failed. Reason: Cannot allocate memory. Error code: 12
  7:39:59.932   170G / 224G  ERROR   General                 (mmapped_reader.hpp        :  52)   mmap(2) failed. Reason: Cannot allocate memory. Error code: 12
  7:39:59.932   170G / 224G  ERROR   General                 (mmapped_reader.hpp        :  52)   mmap(2) failed. Reason: Cannot allocate memory. Error code: 12


== Error ==  system call for: "['/home/ubuntu/SPAdes-3.14.0-Linux/bin/spades-core', '/home/ubuntu/data/public_data/cat/lilbub/spades_output/K55/configs/config.info', '/home/ubuntu/data/public_data/cat/lilbub/spades_output/K55/configs/careful_mode.info']" finished abnormally, OS return value: 255

======= SPAdes pipeline finished abnormally and WITH WARNINGS!

=== Error correction and assembling warnings:
 * 3:47:53.379    80G / 216G  WARN    General                 (kmer_coverage_model.cpp   : 218)   Too many erroneous kmers, the estimates might be unreliable
 * 0:00:00.002    10M / 13M   WARN   StageManager             (stage.cpp                 : 150)   No saved checkpoint
======= Warnings saved to /home/ubuntu/data/public_data/cat/lilbub/spades_output/warnings.log

=== ERRORs:
 * 7:39:59.932   170G / 224G  ERROR   General                 (mmapped_reader.hpp        :  52)   mmap(2) failed. Reason: Cannot allocate memory. Error code: 12
 * system call for: "['/home/ubuntu/SPAdes-3.14.0-Linux/bin/spades-core', '/home/ubuntu/data/public_data/cat/lilbub/spades_output/K55/configs/config.info', '/home/ubuntu/data/public_data/cat/lilbub/spades_output/K55/configs/careful_mode.info']" finished abnormally, OS return value: 255

In case you have troubles running SPAdes, you can write to spades.support@cab.spbu.ru
or report an issue on our GitHub repository github.com/ablab/spades
Please provide us with params.txt and spades.log files from the output directory.

SPAdes log can be found here: /home/ubuntu/data/public_data/cat/lilbub/spades_output/spades.log

Thank you for using SPAdes!
ubuntu@ip-172-31-0-124:~/data/public_data/cat/lilbub$ Connection to ec2-3-8-234-202.eu-west-2.compute.amazonaws.com closed by remote host.
```

Since I have the `spades_output` folder saved with the K21 and K33 steps completed, should I start an instance with more memory and rekick with `--continue`? Any recommendations on how much memory? Thanks in advance.

**Comment by asl** [comment]
Well, this is very expected as you skipped the read error correction. In such case you'd perform quality trimming prior to the assembly.

**Comment by avilella** [comment]
Oh, I see. Is there a better parameter set for this? I want to get a graph
file out of it (.fastg, .gfa) that I can then use to search cDNAs against.

Since this is an Illumina 2x150bp ~30x dataset (Insert size about 550bp), I
don't expect much contiguity from the assembly given the input data, but if
I can get a graph file out of it, then searching cDNAs against it is the
main goal.

Looking forward to your reply,

**Comment by asl** [comment]
Well, if your graph will be discontinuous then you won't obtain much from it, frankly speaking.

So, just stick to defaults

PS: PathRacer and SPAligner could help you with graph investigation

**Comment by avilella** [comment]
Thanks, I already had SPAligner in mind to try.

What do you recommend for this dataset given I only have ~30x? I am
currently following the section in the documentation:
http://cab.spbu.ru/files/release3.12.0/manual.html#sec3.4

There are other Illumina datasets out there, but they are from other
individuals, so I am assuming they will contain individual/breed SNPs.

**Comment by avilella** [comment]
Just for completeness, I've rekicked with `--careful` but without `--only_assemble` on an AWS instance with only 16Gb of RAM but with about 1000Gb of Linux swap of NVMe disk. This instance is about 20x cheaper than the 512Gb RAM instance (largemem) which had an out of memory issue. So far it seems to be slower than the largemem instance, but less than 20x slower:

```
== Running: /home/ubuntu/SPAdes-3.14.0-Linux/bin/spades-hammer /data/public_data/cat/lilbub/spades_output/corrected/configs/config.info

  0:00:00.000     3M / 13M   INFO    General                 (main.cpp                  :  75)   Starting BayesHammer, built from refs/heads/spades_3.14.0, git revision c831f9be30a4364383cd4ebe5b78cdfcaad1acc9
  0:00:00.002     3M / 13M   INFO    General                 (main.cpp                  :  76)   Loading config from /data/public_data/cat/lilbub/spades_output/corrected/configs/config.info
  0:00:00.005     3M / 13M   INFO    General                 (main.cpp                  :  78)   Maximum # of threads to use (adjusted due to OMP capabilities): 2
  0:00:00.005     3M / 13M   INFO    General                 (memory_limit.cpp          :  49)   Memory limit set to 1000 Gb
  0:00:00.005     3M / 13M   INFO    General                 (main.cpp                  :  86)   Trying to determine PHRED offset
  0:00:00.005     4M / 13M   INFO    General                 (main.cpp                  :  92)   Determined value is 33
  0:00:00.005     4M / 13M   INFO    General                 (hammer_tools.cpp          :  38)   Hamming graph threshold tau=1, k=21, subkmer positions = [ 0 10 ]
  0:00:00.005     4M / 13M   INFO    General                 (main.cpp                  : 113)   Size of aux. kmer data 24 bytes
     === ITERATION 0 begins ===
  0:00:00.006     4M / 13M   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index
  0:00:00.006     4M / 13M   INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 32 files using 2 threads. This might take a while.
  0:00:00.006     4M / 13M   INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 1024
  0:00:00.006     4M / 13M   INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 166.667 Gb
  0:00:00.006     4M / 13M   INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 2097152
  0:00:00.007     2G / 2G    INFO   K-mer Splitting          (kmer_data.cpp             :  97)   Processing /data/public_data/cat/lilbub/SRR8377759_1.fastq.gz
  0:00:13.327     2G / 2G    INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 252578 reads
  0:00:26.156     2G / 2G    INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 505236 reads
  0:00:38.892     2G / 2G    INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 757310 reads
  0:00:51.620     2G / 2G    INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 1009549 reads
  0:01:04.724     2G / 2G    INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 1262016 reads
  0:01:17.565     2G / 2G    INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 1514281 reads
  0:01:56.128     2G / 2G    INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 2270268 reads
  0:03:39.044     2G / 2G    INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 4293087 reads
  0:07:17.757     2G / 2G    INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 8584378 reads
  0:14:25.112     2G / 2G    INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 16957268 reads
  0:28:37.765     2G / 2G    INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 33662962 reads
  0:57:18.704     2G / 2G    INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 67351080 reads
  1:54:25.018     2G / 2G    INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 134450162 reads
  3:48:32.680     2G / 2G    INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 268534674 reads
  4:24:06.294     2G / 2G    INFO   K-mer Splitting          (kmer_data.cpp             :  97)   Processing /data/public_data/cat/lilbub/SRR8377759_2.fastq.gz
  7:37:19.992     2G / 2G    INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 536970530 reads
  8:48:36.398     2G / 2G    INFO   K-mer Splitting          (kmer_data.cpp             : 112)   Total 620724532 reads processed
  8:48:36.426    24M / 546M  INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.
 15:33:17.581     6M / 15G   INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 11285155542 kmers in total.
 15:33:17.587     6M / 15G   INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.
 15:44:43.413     6M / 15G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices
 16:53:11.363     6G / 15G   INFO    General                 (kmer_index_builder.hpp    : 150)   Merging final buckets.
 17:04:38.267     6G / 15G   INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 5232990446 bytes occupied (3.70965 bits per kmer).
 17:04:38.269     6G / 15G   INFO   K-mer Counting           (kmer_data.cpp             : 351)   Arranging kmers in hash map order
```

**Comment by ipetrushin** [comment]
> This instance is about 20x cheaper than the 512Gb RAM instance (largemem) which had an out of memory issue. So far it seems to be slower than the largemem instance, but less than 20x slower:

@avilella If this case still actual for you, I might try to assembly this data on our 1Tb RAM node. Now processing the large environmental metaviral dataset by metaviralSPAdes v.3.15 (hope it won't crush).
